src/cr/cube/measures/pairwise_pvalues.py

```diff
@@ -18,61 +18,67 @@
 
 
 # pylint: disable=too-few-public-methods
-class PairwisePvalues:
+class PairwiseSignificance:
     """Value object providing matrix of pairwise-comparison P-values"""
 
     def __init__(self, slice_, axis=0, weighted=True):
         self._slice = slice_
         self._axis = axis
         self._weighted = weighted
 
-    @lazyproperty
-    def values(self):
-        """Square matrix of pairwise Chi-square along axis, as numpy.ndarray."""
+    @classmethod
+    def pvals(cls, slice_, axis=0, weighted=True):
+        return cls._factory(slice_, axis, weighted)._pvals
+
+    @staticmethod
+    def _calculate_chi_squared(numel, proportions, margin, observed):
+        chisq = np.zeros([numel, numel])
+        for i in xrange(1, numel):
+            for j in xrange(0, numel - 1):
+                chisq[i, j] = chisq[j, i] = np.sum(
+                    np.square(proportions[:, i] - proportions[:, j]) / observed
+                ) / (1 / margin[i] + 1 / margin[j])
+        return chisq
+
+    def _calculate_pvals_from_chi_squared(self, props):
         return self._intersperse_insertions_rows_and_columns(
-            1.0 - WishartCDF(self._pairwise_chisq, self._n_min, self._n_max).values
+            1.0 - WishartCDF(props, self._n_min, self._n_max).values
         )
 
+    @staticmethod
+    def _factory(slice_, axis, weighted):
+        if slice_.dim_types[0] == DT.MR_SUBVAR:
+            return _MrXCatPairwiseSignificance(slice_, axis, weighted)
+        return _CatXCatPairwiseSignificance(slice_, axis, weighted)
+
     @lazyproperty
-    def _categorical_pairwise_chisq(self):
-        """Pairwise comparisons (Chi-Square) along axis, as numpy.ndarray.
+    def _insertions_indices(self):
+        """Return H&S indices of the pairwise comparison dimension."""
+        return self._slice.inserted_hs_indices()[1 - self._axis]
 
-        Returns a square, symmetric matrix of test statistics for the null
-        hypothesis that each vector along *axis* is equal to each other.
-        """
-        chisq = np.zeros([self._numel, self._numel])
-        for i in xrange(1, self._numel):
-            for j in xrange(0, self._numel - 1):
-                chisq[i, j] = chisq[j, i] = np.sum(
-                    np.square(self._proportions[:, i] - self._proportions[:, j])
-                    / self._observed
-                ) / (1 / self._margin[i] + 1 / self._margin[j])
+    def _intersperse_insertions_rows_and_columns(self, pairwise_pvals):
+        """Return pvals matrix with inserted NaN rows and columns, as numpy.ndarray.
 
-        return chisq
+        Insertions (Headers and Subtotals) create offset in calculated pvals, and
```

Comment:
Just a little technical writing tip, always strongly prefer an expression based on the singular form. A statement involving a plural is harder to reason about and is often ambiguous. So in this case, it would start "An insertion (a header or subtotal) creates an offset in the calculated pvals ...", or perhaps "Each insertion (header or subtotal) creates an offset ...".

Only when referring to a characteristic inherent in the _collection as a whole_ would you want to choose plural expression, and even then, I usually find it avoids ambiguity to say something like "The insertions collection ..." as opposed to "Insertions" (so still a singular expression, again avoiding the plural).

Reply:
Makes perfect sense. Thanks for the tip.